core/server/services/auth/session/index.js

```diff
@@ -0,0 +1,130 @@
+    {BadRequestError, UnauthorizedError, InternalServerError} = require('../../../lib/common/errors');
+
+const getOrigin = function getOrigin(req) {
+    const origin = req.get('origin');
+    const referrer = req.get('referrer');
+
+    if (!origin && !referrer) {
+        return null;
+    }
+
+    if (origin) {
+        return origin;
+    }
+
+    try {
+        return new URL(req.get('referrer')).origin;
+    } catch (e) {
+        return null;
+    }
+};
+
+const createSession = function createSession(req, res, next) {
+    if (!req.body) {
+        return next(new UnauthorizedError({
+            message: common.i18n.t('errors.middleware.auth.accessDenied')
+        }));
+    }
+    const origin = getOrigin(req);
+    if (!origin) {
+        return next(new BadRequestError({
+            message: common.i18n.t('errors.middleware.auth.unknownOrigin')
+        }));
+    }
+    const {username, password} = req.body;
+    User.check({
+        email: username,
+        password
+    }).then((user) => {
+        req.session.user_id = user.id;
+        req.session.origin = origin;
+        req.session.user_agent = req.get('user-agent');
+        req.session.ip = req.ip;
+        res.sendStatus(201);
+    }).catch(() => {
+        next(new UnauthorizedError({
+            message: common.i18n.t('errors.middleware.auth.accessDenied')
+        }));
+    });
+};
+
+const destroySession = function destroySession(req, res, next) {
+    req.session.destroy((err) => {
+        if (err) {
+            return next(new InternalServerError());
+        }
+        return res.sendStatus(204);
+    });
+};
+
+const getUser = function getUser(req, res, next) {
+    if (!req.session || !req.session.user_id) {
+        req.user = null;
+        return next();
+    }
+    User.findOne({id: req.session.user_id})
+        .then((user) => {
+            req.user = user;
+            next();
+        }).catch(() => {
+            next(new UnauthorizedError({
+                message: common.i18n.t('errors.middleware.auth.accessDenied')
+            }));
+        });
+};
+
+const ensureUser = function ensureUser(req, res, next) {
+    if (req.user && req.user.id) {
+        return next();
+    }
+    next(new UnauthorizedError({
+        message: common.i18n.t('errors.middleware.auth.accessDenied')
+    }));
+};
+
+const getSession = session({
+    store: new SessionStore(Session),
+    secret: settingsCache.get('session-secret'),
+    resave: false,
+    saveUninitialized: false,
+    cookie: {
+        maxAge: 184 * 24 * 60 * 60 * 1000, // number of days in second half of year
+        httpOnly: true,
+        path: '/ghost',
+        sameSite: 'lax',
+        secure: /^https:/.test(config.get('url'))
+    }
+});
+
+const cookieCsrfProtection = function cookieCsrfProtection(req, res, next) {
+    // uninitialized session
+    if (!req.session.origin) {
+        return next();
+    }
+
+    if (req.session.origin !== getOrigin(req)) {
```

Comment:
How does that work? 🤔 
So we load the session based on the session eid - where is `req.session.origin` coming from?

Reply:
`req.session.origin` gets set upon login. This offers CSRF protection as we only allow the session to be used from the same site that the login was issued from.

Reply:
here: https://github.com/TryGhost/Ghost/pull/9867/files/b846894343d9a88bb3ca1f56502549fabf90e533#diff-205b165c5b6cd2eb37f4f5a9eade5711R48

Reply:
Okay thanks, have overseen `req.session.origin`. Does that mean the session cookie contains the original origin the session was created with? Just trying to understand, because we don't store the origin persistent. And the session can only be extracted with the private session key stored on the server i guess (the secret).

Reply:
The session cookie only contains `session.id` and (psuedo) `sign(session.id, session-secret)`. Everything on the `req.session` object is stored in the `session_data` column in the `sessions` table.

So we do store the origin in persistent storage, and the session data can only be extracted server side.

Reply:
Ah okay :) Thanks